src/transforms/ppmt.jl

```diff
@@ -0,0 +1,171 @@
+    α_unit = (1/norm(α)) .* α .* (norm(α) != 1) + α .* (norm(α) == 1)
+    X = Z * α_unit
+    2 * cdf(Normal(), X) .- 1
+end
+
+function projection_index(Z,α,J)
+    #Generates the projection index I defined in the paper:
+    #Friedman, J. H. (1987). Exploratory Projection Pursuit. Journal of the American Statistical Association, 82(397), 249–266. https://doi.org/10.2307/2289161
+    R = r_transform(Z,α)
+    if J <= 0
+        return "Error: J must be non-negative"
+    elseif J == 1
+        return (3/2) * mean(R)^2
+    else
+        I = (3/2) * mean(R)^2
+        Pⱼ₋₂, Pⱼ₋₁ = ones(size(R)), R
+        for j = 2:J
+            Pⱼ₋₂, Pⱼ₋₁ = Pⱼ₋₁, (1/j) * ((2j-1) * R .* Pⱼ₋₁ - (j-1) * Pⱼ₋₂)
+            I += ((2j+1)/2) * (mean(Pⱼ₋₁))^2
+        end
+    end
+    return I
+end
+
+function max_principal(Z,J)
+    #Finds the canonical (positive) vector eᵢ that maximizes the projection index I
+    _, q = size(Z)
+    Iₘₐₓ, iₘₐₓ = 0, 0
+    for i = 1:q
+        eᵢ = Float64[1 * (j==i) for j=1:q]
+        I = projection_index(Z,eᵢ,J)
+        if I > Iₘₐₓ
+            Iₘₐₓ, iₘₐₓ = I, i
+        end
+    end
+    return Float64[1 *(iₘₐₓ==i) for i = 1:q]
+end
+
+function max_diagonal(Z,J)
+    #Given the maximizer principal direction, finds the diagonal direction that #maximizes I.
+    ᾱ = max_principal(Z,J)
+    l = length(ᾱ)
+    I₀ = projection_index(Z,ᾱ,J)
+    for i in 1:l
+        eᵢ = Float64[1*(j==i) for j=1:l]
+        f₊ = projection_index(Z, (1/√(2+2ᾱ[i]))*(ᾱ + eᵢ), J)
+        f₋ = projection_index(Z, (1/√(2-2ᾱ[i]))*(ᾱ - eᵢ), J) * (ᾱ[i] != 1)
+        f = (f₊>f₋)*f₊ + (f₊<=f₋)*f₋
+        s = (f₊>f₋)*1 + (f₊<=f₋)*(-1)
+        if f > projection_index(Z,ᾱ,J)
+            ᾱ = (1/√(2+2s*ᾱ[i]))*(ᾱ + s*eᵢ)
+        end
+    end
+    if projection_index(Z,ᾱ,J) >= I₀
+        return ᾱ
+    else
+        return max_principal(Z,J)
+    end
+end
+
+function optimizer(Z,J)
+    #Given the maximizer diagonal direction, finds a local maximum for I.
+    α₀ = max_diagonal(Z,J)
+    f(α) = -projection_index(Z,α,J)
+    op = optimize(f,α₀)
+    Optim.minimizer(op)
+end
+
+function structure_removal(Z,α)
+    #Transforms the cdf of a convex combination of r.v. to a standard normal cdf.
+    l = length(α)
+    U = qr(hcat(α,rand(l,l-1)))
+    if norm(diag(U.R))<1e-5 #Considers if the randomly created matrix is singular.
+        return print("Congrats: you have hit the almost impossible chance of drawing $l-1 random l.d. vectors!")
+    end
+    T = Z * U.Q
+    cdf = EmpiricalCDF(T[:,1])
+    sort!(cdf)
+    T[:,1] = [quantile(Normal(), q) for q in cdf(T[:,1]) .- 1e-5]
+    return T*U.Q'
+end
+
+function gaussian_projection_matrix(N,q,J,k)
+    #Gives the array that will be used to stop the loop of structure removal by computing the projection index on k samples of multinomial distribution (size of sample kxn) along q random directions.
+    I = ones(k,q)
+    d = MvNormal(zeros(q), Diagonal(ones(q)))
+    I = Matrix{Float64}(undef,k,q)
```

Comment:
As you will learn with more time, we usually never hard-code types such as Float64 in the code, but instead inherit it from other variables. This is to make sure that we are not changing the precision of the input table provided by the user.